src/NumericalAlgorithms/Interpolation/LinearLeastSquares.cpp

```diff
@@ -0,0 +1,84 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+#include "NumericalAlgorithms/Interpolation/LinearLeastSquares.hpp"
+
+#include <algorithm>
+#include <cstddef>
+#include <gsl/gsl_spline.h>
+#include <memory>
+#include <pup.h>
+#include <pup_stl.h>
+#include <vector>
+
+#include "Utilities/ErrorHandling/Assert.hpp"
+
+namespace intrp {
+template <size_t Order>
+LinearLeastSquares<Order>::LinearLeastSquares(
+    const std::vector<double>& x_values, const std::vector<double>& y_values)
+    : x_values_(std::move(x_values)), y_values_(std::move(y_values)) {
+  ASSERT(x_values_.size() == y_values_.size(),
+         "The x-value and y-value vectors must be of the same length, but "
+         "received x-value of size: "
+             << x_values_.size()
+             << " and y-value of size: " << y_values_.size());
+  ASSERT(std::is_sorted(x_values_.begin(), x_values_.end()),
+         "The x-values must be sorted.");
+  const size_t observations = x_values.size();
+  gsl_matrix *X, *cov;
+  gsl_vector *y, *c;
+
+  X = gsl_matrix_alloc(observations, Order + 1);
+  y = gsl_vector_alloc(observations);
+  c = gsl_vector_alloc(Order + 1);
+  cov = gsl_matrix_alloc(Order + 1, Order + 1);
+  for (size_t i = 0; i < observations; i++) {
+    for (size_t j = 0; j < Order + 1; j++) {
+      gsl_matrix_set(X, i, j, pow(x_values[i], j));
+    }
+    gsl_vector_set(y, i, y_values[i]);
+  }
+
+  gsl_multifit_linear_workspace* work =
+      gsl_multifit_linear_alloc(observations, Order + 1);
+  gsl_multifit_linear(X, y, c, cov, &chisq_, work);
+  gsl_multifit_linear_free(work);
+
+  for (size_t i = 0; i < Order + 1; i++) {
+    c_values_[i] = gsl_vector_get(c, i);
+  }
+
+  gsl_matrix_free(X);
+  gsl_vector_free(y);
+  gsl_vector_free(c);
+  gsl_matrix_free(cov);
+}
+
+template <size_t Order>
+double LinearLeastSquares<Order>::operator()(
```

Comment:
There are a bunch of lines that apparently aren't tested. Could you please add tests for them?